File: src/main/java/com/gg/plugins/mongo/service/MongoService.java
/*
 * Copyright (c) 2022. Geetesh Gupta
 */

package com.gg.plugins.mongo.service;

import com.gg.plugins.mongo.config.ConfigurationException;
import com.gg.plugins.mongo.config.ServerConfiguration;
import com.gg.plugins.mongo.config.SshTunnelingConfiguration;
import com.gg.plugins.mongo.config.ssh.SshConnection;
import com.gg.plugins.mongo.model.*;
import com.intellij.openapi.project.Project;
import com.mongodb.*;
import com.mongodb.client.AggregateIterable;
import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoIterable;
import com.mongodb.client.model.FindOneAndReplaceOptions;
import org.apache.commons.lang.StringUtils;
import org.bson.Document;

import java.util.*;

public class MongoService {

	public static final String DEAFULT_AUTHENTICATION_DATABASE = "admin";

	private static final String DEFAULT_TUNNEL_LOCAL_HOST = "localhost";

	private static final int DEFAULT_TUNNEL_LOCAL_PORT = 9080;

	private static final Document EMPTY_DOC = new Document();

	private final List<MongoServer> mongoServers = new LinkedList<>();

	public static MongoService getInstance(Project project) {
		return project.getService(MongoService.class);
	}

	public void connect(final ServerConfiguration configuration) {
		Task task = mongoClient -> {
			String userDatabase = configuration.getUserDatabase();
			String databaseName = StringUtils.isNotEmpty(userDatabase) ? userDatabase : "test";

			mongoClient.getDatabase(databaseName).listCollectionNames().first();
		};

		executeTask(configuration, task);
	}

	private void executeTask(ServerConfiguration configuration, Task perform) {
		if (SshTunnelingConfiguration.isEmpty(configuration.getSshTunnelingConfiguration())) {
			execute(configuration, perform);
		} else {
			try (SshConnection ignored = SshConnection.create(configuration)) {
				execute(configuration, perform);
			}
		}
	}

	private void execute(ServerConfiguration configuration, Task task) {
		try (MongoClient mongoClient = createMongoClient(configuration)) {
			task.run(mongoClient);
		} catch (MongoException ex) {
			throw new ConfigurationException(ex);
		}
	}

	private MongoClient createMongoClient(ServerConfiguration configuration) {
		List<String> serverUrls = configuration.getServerUrls();
		if (serverUrls.isEmpty()) {
			throw new ConfigurationException("Server host is not set");
		}

		List<ServerAddress> serverAddresses = new LinkedList<>();
		if (SshTunnelingConfiguration.isEmpty(configuration.getSshTunnelingConfiguration())) {
			for (String serverUrl : serverUrls) {
				ServerConfiguration.HostAndPort hostAndPort = ServerConfiguration.extractHostAndPort(serverUrl);
				serverAddresses.add(new ServerAddress(hostAndPort.host, hostAndPort.port));
			}
		} else {
			serverAddresses.add(new ServerAddress(DEFAULT_TUNNEL_LOCAL_HOST, DEFAULT_TUNNEL_LOCAL_PORT));
		}

		MongoClientOptions options = MongoClientOptions.builder()
		                                               .sslEnabled(configuration.isSslConnection())
		                                               .readPreference(configuration.getReadPreference())
		                                               .codecRegistry(MongoClient.getDefaultCodecRegistry())
		                                               .build();

		if (StringUtils.isEmpty(configuration.getUsername())) {
			return new MongoClient(serverAddresses, options);
		} else {
			MongoCredential credential = getMongoCredential(configuration);
			return new MongoClient(serverAddresses, credential, options);
		}
	}

	private MongoCredential getMongoCredential(ServerConfiguration configuration) {
		AuthenticationMechanism authenticationMechanism = configuration.getAuthenticationMechanism();
		if (authenticationMechanism == null) {
			return MongoCredential.createPlainCredential(configuration.getUsername(),
					getAuthenticationDatabase(configuration),
					configuration.getPassword().toCharArray());
		} else {
			if (AuthenticationMechanism.SCRAM_SHA_1.equals(authenticationMechanism)) {
				return MongoCredential.createScramSha1Credential(configuration.getUsername(),
						getAuthenticationDatabase(configuration),
						configuration.getPassword().toCharArray());
			}
		}

		throw new IllegalArgumentException("Unsupported authentication macanism: " + authenticationMechanism);
	}

	private static String getAuthenticationDatabase(ServerConfiguration configuration) {
		String authenticationDatabase = configuration.getAuthenticationDatabase();
		return StringUtils.isEmpty(authenticationDatabase) ? DEAFULT_AUTHENTICATION_DATABASE : authenticationDatabase;
	}

	public void cleanUpServers() {
		mongoServers.clear();
	}

	public void registerServer(MongoServer mongoServer) {
		mongoServers.add(mongoServer);
	}

	public List<MongoServer> getServers() {
		return mongoServers;
	}

	public Set<MongoDatabase> loadDatabases(MongoServer mongoServer, ServerConfiguration configuration) {
		final Set<MongoDatabase> mongoDatabases = new TreeSet<>();
		TaskWithReturnedObject<Set<MongoDatabase>> perform = mongoClient -> {
			String userDatabase = configuration.getUserDatabase();

			if (StringUtils.isNotEmpty(userDatabase)) {
				com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(userDatabase);
				MongoDatabase mongoDatabase = new MongoDatabase(database.getName(), mongoServer);
				mongoDatabases.add(createMongoDatabaseAndItsCollections(mongoDatabase, database));
			} else {
				MongoIterable<String> databaseNames = mongoClient.listDatabaseNames();
				for (String databaseName : databaseNames) {
					com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(databaseName);
					MongoDatabase mongoDatabase = new MongoDatabase(database.getName(), mongoServer);
					mongoDatabases.add(createMongoDatabaseAndItsCollections(mongoDatabase, database));
				}
			}
			return mongoDatabases;
		};

		return executeTask(configuration, perform);
	}

	private MongoDatabase createMongoDatabaseAndItsCollections(MongoDatabase mongoDatabase,
			com.mongodb.client.MongoDatabase database) {
		MongoIterable<String> collectionNames = database.listCollectionNames();
		for (String collectionName : collectionNames) {
			mongoDatabase.addCollection(new MongoCollection(collectionName, mongoDatabase));
		}
		return mongoDatabase;
	}

	private <T> T executeTask(ServerConfiguration configuration, TaskWithReturnedObject<T> perform) {
		if (SshTunnelingConfiguration.isEmpty(configuration.getSshTunnelingConfiguration())) {
			return execute(configuration, perform);
		} else {
			try (SshConnection ignored = SshConnection.create(configuration)) {
				return execute(configuration, perform);
			}
		}
	}

	private <T> T execute(ServerConfiguration configuration, TaskWithReturnedObject<T> perform) {
		try (MongoClient mongo = createMongoClient(configuration)) {
			return perform.run(mongo);
		} catch (MongoException mongoEx) {
			throw new ConfigurationException(mongoEx);
		}
	}

	public MongoDatabase loadDatabase(MongoDatabase mongoDatabase, ServerConfiguration configuration) {
		TaskWithReturnedObject<MongoDatabase> perform = mongoClient -> {
			com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(mongoDatabase.getName());
			MongoDatabase mongoDatabaseNew = new MongoDatabase(database.getName(), mongoDatabase.getParentServer());
			return createMongoDatabaseAndItsCollections(mongoDatabaseNew, database);
		};

		return executeTask(configuration, perform);
	}

	public MongoCollectionResult findMongoDocuments(ServerConfiguration configuration,
			final MongoCollection mongoCollection,
			final MongoQueryOptions mongoQueryOptions) {
		TaskWithReturnedObject<MongoCollectionResult> task = mongoClient -> {
			MongoDatabase mongoDatabase = mongoCollection.getParentDatabase();

			com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(mongoDatabase.getName());
			com.mongodb.client.MongoCollection<Document> collection =
					database.getCollection(mongoCollection.getName());

			MongoCollectionResult mongoCollectionResult = new MongoCollectionResult(mongoCollection.getName());
			if (mongoQueryOptions.isAggregate()) {
				return aggregate(mongoQueryOptions, mongoCollectionResult, collection);
			}

			return find(mongoQueryOptions, mongoCollectionResult, collection);
		};

		return executeTask(configuration, task);
	}

	private MongoCollectionResult aggregate(MongoQueryOptions mongoQueryOptions,
			MongoCollectionResult mongoCollectionResult,
			com.mongodb.client.MongoCollection<Document> collection) {
		AggregateIterable<Document> aggregate = collection.aggregate(mongoQueryOptions.getOperations());
		int index = 0;
		Iterator<Document> iterator = aggregate.iterator();
		while (iterator.hasNext() && index < mongoQueryOptions.getResultLimit()) {
			mongoCollectionResult.add(iterator.next());
		}
		return mongoCollectionResult;
	}

	private MongoCollectionResult find(MongoQueryOptions mongoQueryOptions,
			final MongoCollectionResult mongoCollectionResult,
			com.mongodb.client.MongoCollection<Document> collection) {
		Document filter = mongoQueryOptions.getFilter();
		Document projection = mongoQueryOptions.getProjection();
		Document sort = mongoQueryOptions.getSort();

		FindIterable<Document> cursor = collection.find(filter);
		if (!MongoQueryOptions.EMPTY_DOCUMENT.equals(projection)) {
			cursor.projection(projection);
		}

		if (!MongoQueryOptions.EMPTY_DOCUMENT.equals(sort)) {
			cursor.sort(sort);
		}

		int resultLimit = mongoQueryOptions.getResultLimit();
		if (resultLimit > 0) {
			cursor.limit(resultLimit);
		}

		cursor.forEach((Block<Document>) mongoCollectionResult::add);

		return mongoCollectionResult;
	}

	public Document findMongoDocument(ServerConfiguration configuration,
			final MongoCollection mongoCollection,
			final Object _id) {
		TaskWithReturnedObject<Document> task = mongoClient -> {
			MongoDatabase mongoDatabase = mongoCollection.getParentDatabase();
			com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(mongoDatabase.getName());
			com.mongodb.client.MongoCollection<Document> collection =
					database.getCollection(mongoCollection.getName());
			FindIterable<Document> foundDocuments = collection.find(new BasicDBObject("_id", _id));

			return foundDocuments.first();
		};

		return executeTask(configuration, task);
	}

	public void update(ServerConfiguration configuration,
			final MongoCollection mongoCollection,
			final Document mongoDocument) {
		Task task = mongoClient -> {
			MongoDatabase mongoDatabase = mongoCollection.getParentDatabase();
			com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(mongoDatabase.getName());
			com.mongodb.client.MongoCollection<Document> collection =
					database.getCollection(mongoCollection.getName());

			if (!mongoDocument.containsKey("_id")) {
				collection.insertOne(mongoDocument);
			} else {
				collection.findOneAndReplace(new Document("_id", mongoDocument.get("_id")),
						mongoDocument,
						new FindOneAndReplaceOptions().upsert(true));
			}
		};

		executeTask(configuration, task);
	}

	public void delete(ServerConfiguration configuration, final MongoCollection mongoCollection, final Object _id) {
		Task task = mongoClient -> {
			MongoDatabase mongoDatabase = mongoCollection.getParentDatabase();

			com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(mongoDatabase.getName());
			com.mongodb.client.MongoCollection<Document> collection =
					database.getCollection(mongoCollection.getName());

			collection.deleteOne(new Document("_id", _id));
		};

		executeTask(configuration, task);
	}

	public void removeCollection(ServerConfiguration configuration, final MongoCollection mongoCollection) {
		Task task = mongoClient -> {
			MongoDatabase mongoDatabase = mongoCollection.getParentDatabase();

			com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(mongoDatabase.getName());
			com.mongodb.client.MongoCollection<Document> collection =
					database.getCollection(mongoCollection.getName());

			collection.drop();
		};
		executeTask(configuration, task);
	}

	public void removeDatabase(ServerConfiguration configuration, final MongoDatabase selectedDatabase) {
		Task task = mongoClient -> mongoClient.dropDatabase(selectedDatabase.getName());

		executeTask(configuration, task);
	}

	public void importData(ServerConfiguration configuration,
			MongoCollection mongoCollection,
			List<Document> mongoDocuments,
			boolean replaceAllDocuments) {
		Task task = mongoClient -> {
			MongoDatabase mongoDatabase = mongoCollection.getParentDatabase();
			com.mongodb.client.MongoDatabase database = mongoClient.getDatabase(mongoDatabase.getName());
			com.mongodb.client.MongoCollection<Document> collection =
					database.getCollection(mongoCollection.getName());

			if (replaceAllDocuments) {
				collection.deleteMany(EMPTY_DOC);
			}

			collection.insertMany(mongoDocuments);
		};

		executeTask(configuration, task);
	}

	private interface Task {

		void run(MongoClient mongoClient);
	}

	private interface TaskWithReturnedObject<T> {

		T run(MongoClient mongoClient);
	}
}

File: src/main/kotlin/com/gg/plugins/mongo/service/MongoService.kt
/*
 * Copyright (c) 2018 David Boissier.
 * Modifications Copyright (c) 2022 Geetesh Gupta.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.gg.plugins.mongo.service

import com.gg.plugins.mongo.config.ConfigurationException
import com.gg.plugins.mongo.config.ServerConfiguration
import com.gg.plugins.mongo.config.SshTunnelingConfiguration
import com.gg.plugins.mongo.config.ssh.SshConnection
import com.gg.plugins.mongo.model.*
import com.intellij.openapi.project.Project
import com.mongodb.AuthenticationMechanism
import com.mongodb.BasicDBObject
import com.mongodb.MongoClient
import com.mongodb.MongoClientOptions
import com.mongodb.MongoCredential
import com.mongodb.MongoException
import com.mongodb.ServerAddress
import com.mongodb.client.MongoCollection
import com.mongodb.client.model.FindOneAndReplaceOptions
import org.apache.commons.lang.StringUtils
import org.bson.Document
import org.bson.types.ObjectId
import java.util.LinkedList
import java.util.TreeSet
import java.util.function.Consumer

class MongoService {
    private val mongoServers: MutableList<MongoServer?> = LinkedList()
    fun connect(configuration: ServerConfiguration) {
        val task: Task = object : Task {
            override fun run(mongoClient: MongoClient) {
                val userDatabase = configuration.userDatabase
                val databaseName = if (userDatabase.isNullOrEmpty()) "test" else userDatabase
                mongoClient.getDatabase(databaseName).listCollectionNames().first()
            }
        }
        executeTask(configuration, task)
    }

    private fun executeTask(configuration: ServerConfiguration, perform: Task) {
        if (SshTunnelingConfiguration.isEmpty(configuration.sshTunnelingConfiguration)) {
            execute(configuration, perform)
        } else {
            SshConnection.create(configuration).use { execute(configuration, perform) }
        }
    }

    private fun execute(configuration: ServerConfiguration, task: Task) {
        try {
            createMongoClient(configuration).use { mongoClient -> task.run(mongoClient) }
        } catch (ex: MongoException) {
            throw ConfigurationException(ex)
        }
    }

    private fun createMongoClient(configuration: ServerConfiguration): MongoClient {
        val serverUrls = configuration.serverUrls
        if (serverUrls.isEmpty()) {
            throw ConfigurationException("Server host is not set")
        }
        val serverAddresses: MutableList<ServerAddress> = LinkedList()
        if (SshTunnelingConfiguration.isEmpty(configuration.sshTunnelingConfiguration)) {
            for (serverUrl in serverUrls) {
                val hostAndPort: ServerConfiguration.HostAndPort = ServerConfiguration.extractHostAndPort(
                    serverUrl
                )
                serverAddresses.add(ServerAddress(hostAndPort.host, hostAndPort.port))
            }
        } else {
            serverAddresses.add(ServerAddress(DEFAULT_TUNNEL_LOCAL_HOST, DEFAULT_TUNNEL_LOCAL_PORT))
        }
        val options = MongoClientOptions.builder()
            .sslEnabled(configuration.isSslConnection)
            .readPreference(configuration.readPreference)
            .codecRegistry(MongoClient.getDefaultCodecRegistry())
            .build()
        return if (StringUtils.isEmpty(configuration.username)) {
            MongoClient(serverAddresses, options)
        } else {
            val credential = getMongoCredential(configuration)
            MongoClient(serverAddresses, credential, options)
        }
    }

    private fun getMongoCredential(configuration: ServerConfiguration): MongoCredential {
        val authenticationMechanism = configuration.authenticationMechanism
        if (authenticationMechanism == null) {
            return MongoCredential.createPlainCredential(
                configuration.username!!,
                getAuthenticationDatabase(configuration),
                configuration.password!!.toCharArray()
            )
        } else {
            if (AuthenticationMechanism.SCRAM_SHA_1 == authenticationMechanism) {
                return MongoCredential.createScramSha1Credential(
                    configuration.username!!,
                    getAuthenticationDatabase(configuration),
                    configuration.password!!.toCharArray()
                )
            }
        }
        throw IllegalArgumentException("Unsupported authentication mechanism: $authenticationMechanism")
    }

    fun cleanUpServers() {
        mongoServers.clear()
    }

    fun registerServer(mongoServer: MongoServer?) {
        mongoServers.add(mongoServer)
    }

    val servers: List<MongoServer?>
        get() = mongoServers

    fun loadDatabases(mongoServer: MongoServer, configuration: ServerConfiguration): MutableSet<MongoDatabase> {
        val mongoDatabases: MutableSet<MongoDatabase> = TreeSet()
        val perform: TaskWithReturnedObject<MutableSet<MongoDatabase>> =
            object : TaskWithReturnedObject<MutableSet<MongoDatabase>> {
                override fun run(mongoClient: MongoClient): MutableSet<MongoDatabase> {
                    val userDatabase = configuration.userDatabase
                    when {
                        userDatabase.isNullOrEmpty() -> {
                            val databaseNames = mongoClient.listDatabaseNames()
                            for (databaseName in databaseNames) {
                                val database = mongoClient.getDatabase(databaseName)
                                val mongoDatabase = MongoDatabase(database.name, mongoServer)
                                mongoDatabases.add(mongoDatabase)
                            }
                        }

                        else -> {
                            val database = mongoClient.getDatabase(userDatabase)
                            val mongoDatabase = MongoDatabase(database.name, mongoServer)
                            mongoDatabases.add(mongoDatabase)
                        }
                    }
                    return mongoDatabases
                }
            }
        return executeTask(configuration, perform)
    }

    private fun <T> executeTask(configuration: ServerConfiguration, perform: TaskWithReturnedObject<T>): T {
        if (SshTunnelingConfiguration.isEmpty(configuration.sshTunnelingConfiguration)) {
            return execute(configuration, perform)
        } else {
            SshConnection.create(configuration).use { return execute(configuration, perform) }
        }
    }

    private fun <T> execute(configuration: ServerConfiguration, perform: TaskWithReturnedObject<T>): T {
        try {
            createMongoClient(configuration).use { mongo -> return perform.run(mongo) }
        } catch (mongoEx: MongoException) {
            throw ConfigurationException(mongoEx)
        }
    }

    fun loadCollections(
        mongoDatabase: MongoDatabase,
        configuration: ServerConfiguration
    ): Set<com.gg.plugins.mongo.model.MongoCollection> {
        val collections: MutableSet<com.gg.plugins.mongo.model.MongoCollection> = TreeSet()
        val perform: TaskWithReturnedObject<Set<com.gg.plugins.mongo.model.MongoCollection>> =
            object : TaskWithReturnedObject<Set<com.gg.plugins.mongo.model.MongoCollection>> {
                override fun run(mongoClient: MongoClient): Set<com.gg.plugins.mongo.model.MongoCollection> {
                    val database = mongoClient.getDatabase(mongoDatabase.name)
                    val collectionNames = database.listCollectionNames()
                    for (collectionName in collectionNames) {
                        collections.add(MongoCollection(collectionName, mongoDatabase))
                    }
                    return collections
                }
            }
        return executeTask(configuration, perform)
    }

    fun findMongoDocuments(
        configuration: ServerConfiguration,
        mongoCollection: com.gg.plugins.mongo.model.MongoCollection,
        mongoQueryOptions: MongoQueryOptions?
    ): MongoCollectionResult {
        val task: TaskWithReturnedObject<MongoCollectionResult> =
            object : TaskWithReturnedObject<MongoCollectionResult> {
                override fun run(mongoClient: MongoClient): MongoCollectionResult {
                    val mongoDatabase = mongoCollection.parentDatabase
                    val database = mongoClient.getDatabase(mongoDatabase.name)
                    val collection = database.getCollection(mongoCollection.name)
                    val mongoCollectionResult = MongoCollectionResult(mongoCollection.name)
                    if (mongoQueryOptions!!.isAggregate()) {
                        return aggregate(mongoQueryOptions, mongoCollectionResult, collection)
                    }
                    return find(mongoQueryOptions, mongoCollectionResult, collection)
                }
            }
        return executeTask(configuration, task)
    }

    private fun aggregate(
        mongoQueryOptions: MongoQueryOptions,
        mongoCollectionResult: MongoCollectionResult,
        collection: MongoCollection<Document>
    ): MongoCollectionResult {
        val aggregate = collection.aggregate(mongoQueryOptions.operations)
        val index = 0
        val iterator: Iterator<Document> = aggregate.iterator()
        while (iterator.hasNext() && index < mongoQueryOptions.resultLimit) {
            mongoCollectionResult.add(iterator.next())
        }
        return mongoCollectionResult
    }

    private fun find(
        mongoQueryOptions: MongoQueryOptions,
        mongoCollectionResult: MongoCollectionResult,
        collection: MongoCollection<Document>
    ): MongoCollectionResult {
        val filter = mongoQueryOptions.filter
        val projection = mongoQueryOptions.projection
        val sort = mongoQueryOptions.sort
        val cursor = collection.find(filter)
        if (MongoQueryOptions.EMPTY_DOCUMENT != projection) {
            cursor.projection(projection)
        }
        if (MongoQueryOptions.EMPTY_DOCUMENT != sort) {
            cursor.sort(sort)
        }
        val resultLimit = mongoQueryOptions.resultLimit
        if (resultLimit > 0) {
            cursor.limit(resultLimit)
        }
        cursor.forEach(Consumer { document: Document -> mongoCollectionResult.add(document) } as Consumer<in Document>)
        return mongoCollectionResult
    }

    fun findMongoDocument(
        configuration: ServerConfiguration,
        mongoCollection: com.gg.plugins.mongo.model.MongoCollection,
        _id: Any
    ): Document? {
        val task: TaskWithReturnedObject<Document?> = object : TaskWithReturnedObject<Document?> {
            override fun run(mongoClient: MongoClient): Document? {
                val mongoDatabase = mongoCollection.parentDatabase
                val database = mongoClient.getDatabase(mongoDatabase.name)
                val collection = database.getCollection(mongoCollection.name)
                val foundDocuments = collection.find(BasicDBObject("_id", _id))
                return foundDocuments.first()
            }
        }
        return executeTask(configuration, task)
    }

    fun update(
        configuration: ServerConfiguration,
        mongoCollection: com.gg.plugins.mongo.model.MongoCollection,
        mongoDocument: Document?
    ) {
        val task = object : Task {
            override fun run(mongoClient: MongoClient) {
                val mongoDatabase = mongoCollection.parentDatabase
                val database = mongoClient.getDatabase(mongoDatabase.name)
                val collection = database.getCollection(mongoCollection.name)
                if (!mongoDocument!!.containsKey("_id")) {
                    collection.insertOne(mongoDocument)
                } else {
                    collection.findOneAndReplace(
                        Document("_id", mongoDocument["_id"]),
                        mongoDocument,
                        FindOneAndReplaceOptions().upsert(true)
                    )
                }
            }
        }
        executeTask(configuration, task)
    }

    fun delete(
        configuration: ServerConfiguration,
        mongoCollection: com.gg.plugins.mongo.model.MongoCollection,
        _id: ObjectId
    ) {
        val task = object : Task {
            override fun run(mongoClient: MongoClient) {
                val mongoDatabase: MongoDatabase = mongoCollection.parentDatabase
                val database = mongoClient.getDatabase(mongoDatabase.name)
                val collection = database.getCollection(mongoCollection.name)

                collection.deleteOne(Document("_id", _id))
            }
        }
        executeTask(configuration, task)
    }

    fun removeCollection(
        configuration: ServerConfiguration,
        mongoCollection: com.gg.plugins.mongo.model.MongoCollection
    ) {
        val task = object : Task {
            override fun run(mongoClient: MongoClient) {
                val mongoDatabase = mongoCollection.parentDatabase
                val database = mongoClient.getDatabase(mongoDatabase.name)
                val collection = database.getCollection(mongoCollection.name)
                collection.drop()
            }
        }
        executeTask(configuration, task)
    }

    fun removeDatabase(configuration: ServerConfiguration, selectedDatabase: MongoDatabase) {
        val task: Task = object : Task {
            override fun run(mongoClient: MongoClient) {
                mongoClient.dropDatabase(selectedDatabase.name)
            }
        }
        executeTask(configuration, task)
    }

    fun importData(
        configuration: ServerConfiguration,
        mongoCollection: com.gg.plugins.mongo.model.MongoCollection,
        mongoDocuments: List<Document>,
        replaceAllDocuments: Boolean
    ) {
        val task = object : Task {
            override fun run(mongoClient: MongoClient) {
                val mongoDatabase = mongoCollection.parentDatabase
                val database = mongoClient.getDatabase(mongoDatabase.name)
                val collection = database.getCollection(mongoCollection.name)
                if (replaceAllDocuments) {
                    collection.deleteMany(EMPTY_DOC)
                }
                collection.insertMany(mongoDocuments)
            }
        }
        executeTask(configuration, task)
    }

    private interface Task {
        fun run(mongoClient: MongoClient)
    }

    private interface TaskWithReturnedObject<T> {
        fun run(mongoClient: MongoClient): T
    }

    companion object {
        private const val DEFAULT_AUTHENTICATION_DATABASE = "admin"
        private const val DEFAULT_TUNNEL_LOCAL_HOST = "localhost"
        private const val DEFAULT_TUNNEL_LOCAL_PORT = 9080
        private val EMPTY_DOC = Document()
        fun getInstance(project: Project): MongoService {
            return project.getService(MongoService::class.java)
        }

        private fun getAuthenticationDatabase(configuration: ServerConfiguration): String {
            val authenticationDatabase = configuration.authenticationDatabase
            return if (authenticationDatabase.isNullOrEmpty()) DEFAULT_AUTHENTICATION_DATABASE else authenticationDatabase
        }
    }
}